### PR TITLE
chore(scripts/termux_step_install_license): total refactor

### PR DESCRIPTION
**Pull request by TomJo2000**
Part of:
- #20858

This PR is a total refactor of `scripts/build/termux_step_install_license.sh`.
The goal was to eliminate shellcheck warnings and reduce jank.

### CODE REVIEW

**Comment by TomJo2000** [comment]
I'd like some clarification on this check in the current version.
https://github.com/termux/termux-packages/blob/eeca771cecf05365533399c1bfeb264321548d6a/scripts/build/termux_step_install_license.sh#L81-L85

I can't see what it is meant to accomplish.
It seems pretty safe to say that it is intended to loop over all licenses installed for the package.
But the `if` condition doesn't make any sense.
- 1.) The `*` is quoted, and thus not a wildcard.
- 2.) The actually check is if `$LICENSE` is equal to the path where the package licenses are stored.
Which would always be true, if it weren't for the mistakenly quoted `*`,
making it never true unless there is a license called `LICENSE*` which will never be the case.

Since I don't see any way for that check to ever fail, I removed it.
If we do wanna account for all of the license files, I can rewrite the check.
Though that accounting is already done by the rest of the script above.

**Comment by TomJo2000** [comment]
All initial smoke tests seem favorable.
I've rebuilt most of the `root` repo locally with the refactored license installation step changes.

**Comment by Grimler91** [comment]
The check is to verify so that there are files in `$TERMUX_PREFIX/share/doc/$TERMUX_PKG_NAME`. If no licenses were installed for some reason, for example if TERMUX_PKG_LICENSE is empty, then `LICENSE*` will not expand and remain `LICENSE*`, and the check catches that. 

We can probably check this in a more intuitive way, maybe with find.

File: scripts/build/termux_step_install_license.sh [hunk 1]
```diff
@@ -1,87 +1,85 @@
+# shellcheck shell=bash
 termux_step_install_license() {
-	[ "$TERMUX_PKG_METAPACKAGE" = "true" ] && return
+	[[ "$TERMUX_PKG_METAPACKAGE" == 'true' ]] && return
```
Comment (agnostic-apollo) [inline scripts/build/termux_step_install_license.sh:3]: Use `"true"` like everywhere else.

Comment (TomJo2000) [reply scripts/build/termux_step_install_license.sh:3]: They are equivalent.
Since `true` is a static string I used single quotes,
which are not subject to interpolation, unlike double quoted strings.

File: scripts/build/termux_step_install_license.sh [hunk 2]
```diff
@@ -1,87 +1,85 @@
...
-		done < <(echo "$TERMUX_PKG_LICENSE_FILE" | sed "s/,/\n/g")
-	else
-		local TO_LICENSE
+		done <<< "${TERMUX_PKG_LICENSE_FILE//,/$'\n'}"
+	else # If a license file wasn't specified, find the one we need
+		local TO_LICENSE             # link target for generic licenses
+		local COMMON_LICENSE_FILES=( # search list for licenses with copyright information
+		'COPYING' 'Copyright' 'COPYRIGHT' 'Copyright.txt'
+		'licence' 'Licence' 'LICENCE'
+		'license' 'License' 'LICENSE'
```
Comment (agnostic-apollo) [inline scripts/build/termux_step_install_license.sh:35]: Duplicate line.

File: scripts/build/termux_step_install_license.sh [hunk 3]
```diff
@@ -1,87 +1,85 @@
...
 			fi
 			cp -f "${TERMUX_PKG_SRCDIR}/${LICENSE}" "$TARGET"
-		done < <(echo "$TERMUX_PKG_LICENSE_FILE" | sed "s/,/\n/g")
-	else
-		local TO_LICENSE
+		done <<< "${TERMUX_PKG_LICENSE_FILE//,/$'\n'}"
+	else # If a license file wasn't specified, find the one we need
+		local TO_LICENSE             # link target for generic licenses
+		local COMMON_LICENSE_FILES=( # search list for licenses with copyright information
+		'COPYING' 'Copyright' 'COPYRIGHT' 'Copyright.txt'
```
Comment (agnostic-apollo) [inline scripts/build/termux_step_install_license.sh:33]: Add `copyright`.

Comment (TomJo2000) [reply scripts/build/termux_step_install_license.sh:33]: Right, wasn't in the original list.
But probably should be.

File: scripts/build/termux_step_install_license.sh [hunk 4]
```diff
@@ -1,87 +1,85 @@
...
+			if [[ " ${INSTALLED_LICENSES[*]} " == *" $(basename "$LICENSE") "* ]]; then
 				# We have already installed a license file named $(basename $LICENSE) so add a suffix to it
-				TARGET="$TERMUX_PREFIX/share/doc/${TERMUX_PKG_NAME}/$(basename $LICENSE).$COUNTER"
-				COUNTER=$((COUNTER + 1))
+				TARGET="$TERMUX_PREFIX/share/doc/${TERMUX_PKG_NAME}/$(basename "$LICENSE").$((COUNTER++))"
 			else
-				TARGET="$TERMUX_PREFIX/share/doc/${TERMUX_PKG_NAME}/$(basename $LICENSE)"
-				INSTALLED_LICENSES+=("$(basename $LICENSE)")
+				TARGET="$TERMUX_PREFIX/share/doc/${TERMUX_PKG_NAME}/$(basename "$LICENSE")"
+				INSTALLED_LICENSES+=("$(basename "$LICENSE")")
```
Comment (agnostic-apollo) [inline scripts/build/termux_step_install_license.sh:26]: Set `$(basename "$LICENSE")` to a variable instead of it being referenced `4` times.

File: scripts/build/termux_step_install_license.sh [hunk 5]
```diff
@@ -1,87 +1,85 @@
...
+			[[ -z "${LICENSE}" ]] && continue
+
+			# Check that the license file exists in the source files
+			[[ -f "$TERMUX_PKG_SRCDIR/$LICENSE" ]] || {
 				termux_error_exit "$TERMUX_PKG_SRCDIR/$LICENSE does not exist"
-			fi
-			if [[ " ${INSTALLED_LICENSES[@]} " =~ " $(basename $LICENSE) " ]]; then
+			}
+
+			if [[ " ${INSTALLED_LICENSES[*]} " == *" $(basename "$LICENSE") "* ]]; then
```
Comment (agnostic-apollo) [inline scripts/build/termux_step_install_license.sh:21]: Ensure `basename` of license does not contain a space itself.

File: scripts/build/termux_step_install_license.sh (hunk 2 above)
Comment (agnostic-apollo) [reply scripts/build/termux_step_install_license.sh:35]: Ah, sorry.

File: scripts/build/termux_step_install_license.sh (hunk 5 above)
Comment (TomJo2000) [reply scripts/build/termux_step_install_license.sh:21]: That would not matter based on the quoting.

File: scripts/build/termux_step_install_license.sh (hunk 1 above)
Comment (agnostic-apollo) [reply scripts/build/termux_step_install_license.sh:3]: I meant in all other build scripts we always use double quotes and also a single `[`, so best stay consistent. Only python scripts use single quotes.

File: scripts/build/termux_step_install_license.sh (hunk 2 above)
Comment (TomJo2000) [reply scripts/build/termux_step_install_license.sh:35]: Tripped me up too, I can add a comment.

File: scripts/build/termux_step_install_license.sh (hunk 1 above)
Comment (TomJo2000) [reply scripts/build/termux_step_install_license.sh:3]: We also make extensive use of Bash specific features such as arrays.
I don't think we should cling to such POSIX-isms if our build scripts are written for Bash.

Comment (TomJo2000) [reply scripts/build/termux_step_install_license.sh:3]: It's not a hill I'm gonna choose to die on,
but `[[` bash test `]]` has much better defined semantics, condition chaining, additional test operators and regex support.

See the Wooledge Bash Wiki article on conditionals from about this paragraph.
https://mywiki.wooledge.org/BashGuide/TestsAndConditionals#Conditional_Blocks_.28if.2C_test_and_.5B.5B.29:~:text=pitfall%20when%20%5B%20is%20used

<h1><em></em></h1> 

The quoting I can go either way on, I just prefer single quoting static strings (and fwiw `shellharden` agrees with that stance)

**Comment by agnostic-apollo** [comment]
Looks fine otherwise from a quick look.

However, note that debian does not use `LICENSE` as file names, it uses (a single) `copyright`, so would be better to shift to that.

- https://www.debian.org/releases/stable/i386/ch01s08.en.html

Additionally, lot of packages use the machine-readable `debian/copyright` file on debian, including `ruby`, `zsh`, etc. It puts all the license texts under a single file. You can find those packages on the system with `grep -r -l -F "https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/" /usr/share/doc | sort`. I personally don't like the part about putting all the license texts under the same file as that's too verbose and harder to understand, so I break the spec and use `License: [<license_name>](<path_to_license_text_file>)` instead of the  `License: <license_name>\n<license_text>`, hopefully a future spec version adds support for that.

- https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/
- https://lists.debian.org/debian-project/2010/08/msg00269.html
- https://wiki.debian.org/Proposals/CopyrightFormat
- http://ftp.debian.org/debian/pool/main/
- https://askubuntu.com/a/620069

**Comment by TomJo2000** [comment]
Well if you're in the mood to break APIs.
I did briefly think about replacing the `$COUNTER` suffix with the `$LICENSE` identifier.
So it would be `LICENSE.MIT` and `LICENSE.HPND`,
but I dropped that idea because of some of the less filename friendly license idenifiers.
And maintaining API compatibility with the previous version of the license install step.

**Comment by TomJo2000** [comment]
> Looks fine otherwise from a quick look.
> 
> However, note that debian does not use `LICENSE` as file names, it uses (a single) `copyright`, so would be better to shift to that.
> 
> * https://www.debian.org/releases/stable/i386/ch01s08.en.html
> 
> Additionally, lot of packages use the machine-readable `debian/copyright` file on debian, including `ruby`, `zsh`, etc. It puts all the license texts under a single file. You can find those packages on the system with `grep -r -l -F "https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/" /usr/share/doc | sort`. I personally don't like the part about putting all the license texts under the same file as that's too verbose and harder to understand, so I break the spec and use `License: [<license_name>](<path_to_license_text_file>)` instead of the `License: <license_name>\n<license_text>`, hopefully a future spec version adds support for that.
> 
> * https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/
> * https://lists.debian.org/debian-project/2010/08/msg00269.html
> * https://wiki.debian.org/Proposals/CopyrightFormat
> * http://ftp.debian.org/debian/pool/main/
> * https://askubuntu.com/a/620069

I've had a read through the linked resources.
And while I think we might wanna consider that, I think it's out of scope for this PR.
I'm also fairly sure it would necessitate rebuilds of **all** packages, no?

**Comment by agnostic-apollo** [comment]
For checking if license file was added.

```shell
license_files="$(find "$TERMUX_PREFIX/share/doc/$TERMUX_PKG_NAME" -maxdepth 1 -type f -name "LICENSE*")"
if [ -z "$license_files" ]; then
    # error out
fi
```

For mulitple filenames, you can use something like `\( -name "LICENSE*" -o -name "copyright*" \)`.

**Comment by TomJo2000** [comment]
> For checking if license file was added.
> 
> ```shell
> license_files="$(find "$TERMUX_PREFIX/share/doc/$TERMUX_PKG_NAME" -maxdepth 1 -type f -name "LICENSE*")"
> if [ -z "$license_files" ]; then
>     # error out
> fi
> ```
> 
> For mulitple filenames, you can use something like `\( -name "LICENSE*" -o -name "copyright*" \)`.

Should that be a `find -L`?
I guess we don't really care if its a symlink or not, so no?

**Comment by agnostic-apollo** [comment]
I don't suggest using the machine-readable `debian/copyright` format right now, I just wanted to notify that the standard exists, and that it solves quite a few issues, especially in relation to which source files are under what license and the format being machine readable.

However, I do suggest that we use `copyright` and `copyright.<counter>` format right now as that is the standard which wouldn't be hard to follow and would provide consistency with tools processing package license files. Packages can get rebuilt whenever they are.

**Comment by TomJo2000** [comment]
> However, I do suggest that we use `copyright` and `copyright.<counter>` format right now as that is the standard which wouldn't be hard to follow and would provide consistency with tools processing package license files. Packages can get rebuilt whenever they are.

Yep, that's definitely far easier than implementing the whole standard right away.
Done locally now.

**Comment by agnostic-apollo** [comment]
>Should that be a find -L?

Yeah, best use `\( -type f -o -type l \)` instead.

**Comment by agnostic-apollo** [comment]
Thanks.

**Comment by agnostic-apollo** [comment]
>Eventual consistency, eh? :P

Well, if you wanna rebuild all 2000+ packages, who am I to say no, have fun with failed builds and missing source urls :p

**Comment by TomJo2000** [comment]
> Well, if you wanna rebuild all 2000+ packages, who am I to say no, have fun with failed builds and missing source urls :p

CI's been flakey lately anyway.
I sure as hell am not poking it with the "full repo rebuild" stick.

File: scripts/build/termux_step_install_license.sh (hunk 5 above)
Comment (agnostic-apollo) [reply scripts/build/termux_step_install_license.sh:21]: Yeah, I think that would trigger the earlier file not found error due to IFS splitting on space.

Comment (TomJo2000) [reply scripts/build/termux_step_install_license.sh:21]: There is no word splitting happening.
It's quoted.

Comment (agnostic-apollo) [reply scripts/build/termux_step_install_license.sh:21]: But generally, that check needs to be added, otherwise `" foo bar baz "` would match against ` " foo bar "` too.

File: scripts/build/termux_step_install_license.sh (hunk 1 above)
Comment (agnostic-apollo) [reply scripts/build/termux_step_install_license.sh:3]: Oh, its fine using `[[`, I use the same, unless writing posix compatible scripts. that's why I initially just mentioned double quotes. There is no instance of `'true'` check in any bash shell script in the entire repo.

File: scripts/build/termux_step_install_license.sh (hunk 5 above)
Comment (TomJo2000) [reply scripts/build/termux_step_install_license.sh:21]: > But generally, that check needs to be added, otherwise `" foo bar baz "` would match against ` " foo bar "` too.

Nope
image

Comment (agnostic-apollo) [reply scripts/build/termux_step_install_license.sh:21]: >There is no word splitting happening.

I meant in `read`.

Comment (TomJo2000) [reply scripts/build/termux_step_install_license.sh:21]: Except...
That isn't what we have, is it?
image

Now I get what you meant.

Comment (agnostic-apollo) [reply scripts/build/termux_step_install_license.sh:21]: >Nope

You are missing the globs.

`[[ " foo bar baz " == *" foo bar "* ]] && echo true`

Comment (agnostic-apollo) [reply scripts/build/termux_step_install_license.sh:21]: >Now I get what you meant.

Yup. People generally don't add that check for the splitter itself, so can result in false positives.

Comment (TomJo2000) [reply scripts/build/termux_step_install_license.sh:21]: We could sidestep the issue by coverting `INSTALLED_LICENSES=()` to a hashmap, and matching on the keys.
Although that is admittedly slightly cursed.

Comment (agnostic-apollo) [reply scripts/build/termux_step_install_license.sh:21]: Can always check if `$TERMUX_PKG_LICENSE_FILE` contains a space before looping, that should keep current logic. The patch files containing a space are also missed due to similar issues, see https://github.com/termux/termux-packages/commit/2b27086129ef9d89f022bb482daf92e889f4b3da.

Comment (TomJo2000) [reply scripts/build/termux_step_install_license.sh:21]: That would need additional logic compared to a key lookup in an associative array.

Comment (agnostic-apollo) [reply scripts/build/termux_step_install_license.sh:21]: If referring to patch files, can be solved by running `find . -name "*.patch"` and grepping for a space or tab.

Comment (TomJo2000) [reply scripts/build/termux_step_install_license.sh:21]: I was thinking something along the lines of...
```bash
declare -A INSTALLED_LICENSES=()
# [...]
if (( ${#INSTALLED_LICENSES[$LICENSE_FILEPATH]:-} )); then
    # [...]
else
    # yes this is a valid way to assign key value pairs to a hashmap
    INSTALLED_LICENSES+=("${LICENSE_FILEPATH}" 'already installed')
fi
```
I might have understated when I said matching on hashmap keys was a slightly cursed approach.

Comment (agnostic-apollo) [reply scripts/build/termux_step_install_license.sh:21]: This should work if you wanna use associative array.

```shell
unset INSTALLED_LICENSES
local -A INSTALLED_LICENSES=()
local license_filename
while read -r LICENSE; do
    ...
    license_filename="$(basename "$LICENSE")"
    if [[ -n "${INSTALLED_LICENSES["$license_filename"]:-}" ]]; then
    	echo "already installed"
        continue
    fi
   
    INSTALLED_LICENSES["$license_filename"]=1
    # install now

done <<< "${TERMUX_PKG_LICENSE_FILE//,/$'\n'}"
```

Comment (TomJo2000) [reply scripts/build/termux_step_install_license.sh:21]: Okay we're on the same page then.
I think I will use your version, it looks slightly less cursed.

Comment (agnostic-apollo) [reply scripts/build/termux_step_install_license.sh:21]: Lolz, go ahead, and I am pretty sure glob pattern is much more cursed than associative array. This loop also works if you want a normal array at the end by maintaining a separate `local -a` array for it and appending to it if "already added" list doesn't have the element.

Comment (TomJo2000) [reply scripts/build/termux_step_install_license.sh:21]: `poppler-data` should be the perfect "smoke test" for this problem.
https://github.com/termux/termux-packages/blob/9b9e2b87f6c751958b7ef0a546d42756cb96002f/packages/poppler-data/build.sh#L1-L9

**Comment by TomJo2000** [comment]
Alright, that's:
- `copyright.${counter}` as license name
- adding `copyright` to the search list
- turning the license filepath into a variable
- accursed hashmap magic

If you still want the `find` check for making sure all the licenses are present and accounted for I'd appreciate it if you could provide me a cleaned up version of how you want that to work.

**Comment by TomJo2000** [comment]
> Based on above, best use following. It will not match broken symlinks to license files, assuming that's what we want.
> 
> ```shell
> license_files="$(find -L "$TERMUX_PREFIX/share/doc/$TERMUX_PKG_NAME" -maxdepth 1 -type f -name "copyright*")"
> if [ -z "$license_files" ]; then
>     # error out
> fi
> ```

Yep I think that tracks with what we are actually trying to check.

**Comment by agnostic-apollo** [comment]
I think license file symlinks may not be valid during build time, let me check.

**Comment by agnostic-apollo** [comment]
Yup, they wouldn't exist, use following with `\( -type f -o -type l \)`.


```shell
license_files="$(find -L "$TERMUX_PREFIX/share/doc/$TERMUX_PKG_NAME" -maxdepth 1 \( -type f -o -type l \) -name "copyright*")"
if [ -z "$license_files" ]; then
    # error out
fi
```

**Comment by agnostic-apollo** [comment]
By symlinks not being valid, I meant that `termux-licenses` wouldn't exist at `../../../../share/LICENSES/${LICENSE}.txt` at build time and only in `$TERMUX_SCRIPTDIR/packages/termux-licenses/LICENSES/${LICENSE}.txt`. They would be valid on the device though.

**Comment by TomJo2000** [comment]
Going with
```bash
local license_files
license_files="$(find -L "$TERMUX_PREFIX/share/doc/$TERMUX_PKG_NAME" -maxdepth 1 \( -type f -o -type l \) -name "copyright*")"
[[ -n "$license_files" ]] || {
	termux_error_exit "No LICENSE file was installed for $TERMUX_PKG_NAME"
}
```
which is semantically identical.

**Comment by agnostic-apollo** [comment]
>which is semantically identical.

How dare you!

**Comment by TomJo2000** [comment]
> How dare you!

I think the `||` more clearly identifies this as an exception path.

**Comment by agnostic-apollo** [comment]
Yeah, yeah, it's fine.

**Comment by TomJo2000** [comment]
This would only catch all license files missing.
Am I correct in that assumption?

Just wondering if we wanna do something about individual missing licenses.

**Comment by agnostic-apollo** [comment]
Yes, check would pass if at least one license file exists.

If you wanna ensure no missing licenses, in case, a specific loop never ran to fail for `cp`/`ln`, etc, then use

```shell
local license_list
if [[ -n "${TERMUX_PKG_LICENSE_FILE}" ]]; then
    license_list="$TERMUX_PKG_LICENSE_FILE"
else
    license_list="$TERMUX_PKG_LICENSE"
fi

...

license_files="$(find -L "$TERMUX_PREFIX/share/doc/$TERMUX_PKG_NAME" -maxdepth 1 \( -type f -o -type l \) -name "copyright*")"
if [ -z "$license_files" ]; then
    # error out
fi

if [ "$(echo "$license_files" | wc -l)" != "$(echo "$license_list" | sed -E -e 's/[ \t]//g' -e 's/,/\n/g' | wc -l)" ]; then
    # error out
fi
```

**Comment by TomJo2000** [comment]
I think I'll just be satisfied with `cp`'s error message if it can't find the license file in the source repo.

And `ln` is only invoked for "known generic licenses",
so it should be safe to assume those will not be missing.
